modules/replicator/src/integrationTest/java/org/apache/ignite/internal/replicator/ItPlacementDriverReplicaSideTest.java

```diff
@@ -0,0 +1,460 @@
+
+            res.put(nodeName, srvc);
+        }
+
+        return res;
+    }
+
+    /**
+     * Creates a network handler to intercept {@link StopLeaseProlongationMessage}.
+     *
+     * @param handlerService Cluster service to handle message.
+     * @return Network handler.
+     */
+    private NetworkMessageHandler leaseDenyMessageHandler(ClusterService handlerService) {
+        return (msg, sender, correlationId) -> {
+            if (!(msg instanceof PlacementDriverActorMessage)) {
+                return;
+            }
+
+            var handlerNode = handlerService.topologyService().localMember();
+
+            log.info("Lease is denied [replica={}, actor={}, redirect={}]", sender, handlerNode.name(),
+                    ((StopLeaseProlongationMessage) msg).redirectProposal());
+
+            if (denyLeaseHandler != null) {
+                denyLeaseHandler.accept((StopLeaseProlongationMessage) msg, sender, handlerNode.name());
+            }
+        };
+    }
+
+    @Test
+    public void testNotificationToPlacementDriverAboutChangeLeader() throws Exception {
+        Set<String> grpNodes = chooseRandomNodes(3);
+
+        log.info("Replication group is based on {}", grpNodes);
+
+        var raftClientFut = createReplicationGroup(GROUP_ID, grpNodes);
+
+        var raftClient = raftClientFut.get();
+
+        raftClient.refreshLeader().get();
+
+        var leaderNodeName = raftClient.leader().consistentId();
+
+        var newLeaderNodeName = grpNodes.stream().filter(n -> !n.equals(leaderNodeName)).findAny().get();
+
+        log.info("Leader is moving form {} to {}", leaderNodeName, newLeaderNodeName);
+
+        ConcurrentHashMap<String, String> nodesToReceivedDeclineMsg = new ConcurrentHashMap<>();
+
+        denyLeaseHandler = (msg, from, to) -> {
+            nodesToReceivedDeclineMsg.put(to, from);
+        };
+
+        raftClient.transferLeadership(new Peer(newLeaderNodeName)).get();
+
+        var anyNode = randomNode(Set.of());
+
+        log.info("Message sent from {} to {}", anyNode, leaderNodeName);
+
+        var clusterService = clusterServices.get(anyNode);
+
+        new ReplicaService(clusterService.messagingService(), clock).invoke(
+                clusterService.topologyService().getByConsistentId(leaderNodeName),
+                TEST_REPLICA_MESSAGES_FACTORY.primaryReplicaTestRequest()
+                        .groupId(GROUP_ID)
+                        .build()
+        );
+
+        assertTrue(waitForCondition(() -> nodesToReceivedDeclineMsg.size() == placementDriverNodeNames.size(), 10_000));
+
+        for (String nodeName : nodesToReceivedDeclineMsg.keySet()) {
+            assertEquals(leaderNodeName, nodesToReceivedDeclineMsg.get(nodeName));
+
+            assertTrue(placementDriverNodeNames.contains(nodeName));
+        }
+
+        stopReplicationGroup(GROUP_ID, grpNodes);
+    }
+
+    @Test
+    public void testNotificationToPlacementDriverAboutMajorityLoss() throws Exception {
```

Comment:
both tests still are almost the same, maybe create another common method with basic scenario and checks?

Reply:
I do not think it will easy to understand.

Reply:
ok